### PR TITLE
Improve sentence detection using a tokenizer

### LINKED ISSUE #582: Sentence detection is broken in case of parenthesis after period
The sentence detection is broken in case of a parenthesis after a period. A parenthesis after a period should not be invalidated.

### PR DESCRIPTION
**Pull request by atimmer**
This should fix the words bugs that have to do with sentence detection.

To test this: Make sure sentences are detected correctly and check the unit tests if I have missed any other cases.

Fixes #582

### CODE REVIEW

File: js/helpers/html.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+var core = require( "tokenizer2/core" );
+var forEach = require( "lodash/forEach" );
+var memoize = require( "lodash/memoize" );
+
+var tokens = [];
+var htmlBlockTokenizer;
+
+/**
+ * Creates a tokenizer to tokenize HTML into blocks
```
Comment (terw-dan) [inline js/helpers/html.js]: Please add a period after the sentence.

File: js/helpers/html.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+	htmlBlockTokenizer.addRule( commentStartRegex, "comment-start" );
+	htmlBlockTokenizer.addRule( commentEndRegex, "comment-end" );
+
+	htmlBlockTokenizer.addRule( otherElementStartRegex, "other-element-start" );
+	htmlBlockTokenizer.addRule( otherElementEndRegex, "other-element-end" );
+}
+
+/**
+ * Returns whether or not the given element name is a block element
```
Comment (terw-dan) [inline js/helpers/html.js]: Please add a period after the sentence.

File: js/helpers/html.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+ *
+ * @param {string} htmlElementName The name of the HTML element.
+ * @returns {boolean} Whether or not it is a block element.
+ */
+function isBlockElement( htmlElementName ) {
+	return blockElementsRegex.test( htmlElementName );
+}
+
+/**
+ * Returns whether or not the given element name is an inline element
```
Comment (terw-dan) [inline js/helpers/html.js]: See ^^

File: js/helpers/html.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	try {
+		htmlBlockTokenizer.end();
+	} catch ( e ) {
+		return [];
+	}
+
+	forEach( tokens, function( token, i ) {
+		var nextToken = tokens[ i + 1 ];
+
+		switch ( token.type ) {
```
Comment (terw-dan) [inline js/helpers/html.js]: Is there a reason this switch statement doesn't have a default?

File: js/stringProcessing/getSentences.js [hunk 5]
```diff
@@ -1,135 +1,198 @@
...
+var tokens = [];
+var sentenceTokenizer;
 
 /**
- * Checks if the period is followed with a whitespace or < for an html-tag. If not, it is no ending of a sentence.
- *
- * @param {string} text The text to split in sentences.
- * @param {number} index The current index to look for.
- * @returns {boolean} True if it doesn't match a whitespace or < .
+ * Creates a tokenizer to create tokens from a sentence
```
Comment (terw-dan) [inline js/stringProcessing/getSentences.js]: Please add a period at the end of the sentence.

File: js/stringProcessing/getSentences.js [hunk 6]
```diff
@@ -1,135 +1,198 @@
...
+	sentenceTokenizer.addRule( htmlStartRegex, "html-start" );
+	sentenceTokenizer.addRule( htmlEndRegex, "html-end" );
+	sentenceTokenizer.addRule( sentenceDelimiterRegex, "sentence-delimiter" );
+	sentenceTokenizer.addRule( sentenceRegex, "sentence" );
+}
 
 /**
- * Checks if the character in the next sentence is a capital letter or a number. If not, this period is not the end
- * of a sentence.
+ * Returns whether or not a certain character is a capital letter
```
Comment (terw-dan) [inline js/stringProcessing/getSentences.js]: Please add a period at the end of the sentence.

File: js/stringProcessing/getSentences.js [hunk 7]
```diff
@@ -1,135 +1,198 @@
...
+function isBreakTag( htmlTag ) {
+	return /<br/.test( htmlTag );
+}
 
 /**
- * Filters the positions that aren't valid endings to sentences.
- * @param {string} text The text to split in sentences.
- * @param {Array} positions The array with positions of periods in the text.
- * @returns {Array} The filtered positions.
+ * Tokenizes a sentence, assumes that the text has already been split into blocks
```
Comment (terw-dan) [inline js/stringProcessing/getSentences.js]: Please add a period at the end of the sentence.

File: js/stringProcessing/getSentences.js [hunk 8]
```diff
@@ -1,135 +1,198 @@
...
+	sentenceTokenizer.onText( text );
+
+	sentenceTokenizer.end();
+
+	return tokens;
+}
 
 /**
- * Split sentences based on index.
+ * Returns an array of sentences for a given array of tokens, assumes that the text has already been split into blocks
```
Comment (terw-dan) [inline js/stringProcessing/getSentences.js]: Please add a period at the end of the sentence.

**Comment by terw-dan** [comment]
CR Done

File: js/helpers/html.js (hunk 4 above)
Comment (atimmer) [inline js/helpers/html.js]: Yes, tokens we don't recognise should just be discarded.

**Comment by terw-dan** [comment]
CR Done :+1:

**Pull request merged**